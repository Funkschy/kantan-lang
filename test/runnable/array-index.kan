import "io";

extern def malloc(size: usize): *void;

let buf: [4]i8 = "   ";

def main() {
    let arr = malloc(5 * 5 * sizeof string) as **string;
    defer delete arr;

    let a: [2][2]i8 = undefined;
    a[0][0] = 'a';
    io.printf("%c\n", a[0][0]);

    let b: [2]i8 = undefined;
    b[0] = 'b';
    io.printf("%c\n", b[0]);

    for let i = 0; i < 2; i += 1 {
        for let j = 0; j < 2; j += 1 {
            a[i][j] = (i * 2 + j) as i8;
        }
    }
    for let i = 0; i < 2; i += 1 {
        for let j = 0; j < 2; j += 1 {
            io.printf("%d\n", a[i][j] as i32);
        }
    }

    let test = Test { i: undefined };
    test.i[0] = 10;
    test.i[1] = 20;

    let p = &test;
    io.printf("%d\n", (*p).i[1]);

    let arr: [2]*Test = undefined;
    arr[0] = arr[1] = p;

    io.printf("%d\n", (*arr[0]).i[0]);

    let nums: [1][1]i32 = undefined;
    nums[0][0] = 1;
    io.printf("%d\n", nums[0][0]);

    let i = &nums[0][0];
    *i = 2;
    io.printf("%d\n", nums[0][0]);

    buf[1] = 'a';
    io.printf("%s\n", &buf[0]);
}

type Test struct {
    i: [2]i32
}

