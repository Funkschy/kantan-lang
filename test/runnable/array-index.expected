a
b
0
1
2
3
20
10
1
2
 a  
